springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java

```diff
@@ -0,0 +1,202 @@
+        Optional<IexQuote> iexQuote;
+        List<IexQuote> quotes = findAllById(Collections.singletonList(ticker));
+
+        if (quotes.size() == 0) {
+            return Optional.empty();
+        }
+        else if (quotes.size() == 1) {
+            iexQuote = Optional.of(quotes.get(0));
+        }
+        else {
+            throw new DataRetrievalFailureException("Unexpected number of quotes");
+        }
+
+        return iexQuote;
+    }
+
+    /**
+     * Gets quotes from IEX
+     *
+     * @param tickers is a list of tickers
+     * @return a list of IexQuote objects
+     * @throws IllegalArgumentException if any given ticker is invalid or if tickers is empty
+     * @throws DataRetrievalFailureException if the HTTP request failed
+     */
+    @Override
+    public List<IexQuote> findAllById(Iterable<String> tickers) {
+        int tickerCount = 0;
+
+        //Go through tickers and make sure they are proper format
+        for (String ticker : tickers) {
+            if (!ticker.matches("[a-zA-Z]*")) {
+                throw new IllegalArgumentException("Incorrect ticker format");
+            }
+            tickerCount++;
+        }
+
+        //Check if tickers is empty
+        if (tickerCount == 0) {
+            throw new IllegalArgumentException("No tickers specified");
+        }
+
+        //Construct url with all tickers
+        String tickerString = String.join(",", tickers);
+        String url = String.format(IEX_BATCH_URL, tickerString);
+
+        //Execute httpGet then convert the response to JSON
+        Optional<String> response = executeHttpGet(url);
+        JSONObject jsonQbject = new JSONObject(response.get());
+
+        //return a list of quotes after mapping JSON to IexQuote objects
+        List<IexQuote> listOfQuotes = new ArrayList<>();
+        ObjectMapper mapper = new ObjectMapper();
+
+        for (String ticker : tickers) {
+            try {
+                String quoteString = jsonQbject.getJSONObject(ticker).get("quote").toString();
+                IexQuote quote = mapper.readValue(quoteString, IexQuote.class);
+                listOfQuotes.add(quote);
+            }
+            catch (IOException e) {
+                throw new DataRetrievalFailureException("Error converting JSON to quote object.");
+            }
+        }
+        return listOfQuotes;
+    }
+
+    /**
+     * Execute a get and return http entity/body as a string
+     *
+     * @param url resource URL
+     * @return HTTP response body or Optional.empty for 404 response
+     * @throws DataRetrievalFailureException if HTTP failed or status code is unexpected
+     */
+    private Optional<String> executeHttpGet(String url) {
+        HttpClient httpClient = getHttpClient();
+        HttpGet httpGet = new HttpGet(url);
+
+        try {
+            HttpResponse response = httpClient.execute(httpGet);
+            int status = response.getStatusLine().getStatusCode();
+
+            if (status == 200) {
```

Comment:
It's more intuitive to do this way

           if (status != 200) {
                throw new DataRetrievalFailureException("Unexpected status code: " + status);
            }
                HttpEntity responseBody = response.getEntity();
                return Optional.of(EntityUtils.toString(responseBody));

Reply:
this is because you might have a lot of lines below the if-block..